### PR TITLE
Enable nuget packages for on device training

### PR DESCRIPTION
**Pull request by askhade**
### Description
This PR enables building nuget packages locally for on device training using --build_nuget arg.
This PR also enables the C# bindings by default in the managed package. If a user triggers any training apis when the native binary is not built for training, an exception with message "Training is disabled in the current build. Please build ONNXRuntime from source with the build flags enable_training and enable_training_on_device. " is thrown. 

Build command for creating nuget packes for on device training:
build.bat --enable_training --enable_training_on_device --build_nuget 

2 Nuget packages are built
1. Microsoft.ML.OnnxRuntime.Managed
2. Microsoft.ML.OnnxRuntime.Training OR Microsoft.ML.OnnxRuntime.Training.Gpu

### CODE REVIEW

File: csharp/OnnxRuntime.CSharp.proj [hunk 1]
```diff
@@ -105,15 +105,15 @@ CMake creates a target to this project
 
     <Copy
     SourceFiles="$(NativeBuildOutputDirAbs)\$(OrtPackageId).$(PackageVersion).nupkg"
-    DestinationFolder="$(NativeBuildOutputDirAbs)\nuget-artifacts"
+    DestinationFolder="$(NativeBuildOutputDirAbs)\nuget-local-artifacts"
     />
```
Comment (skottmckay) [inline csharp/OnnxRuntime.CSharp.proj:109]: There are some CI files that I think use the nuget-artifacts directory. Do those need updating?

\tools\ci_build\github\azure-pipelines\templates\c-api-cpu.yml
\tools\ci_build\github\windows\extract_nuget_files.ps1
\tools\ci_build\github\window\extract_nuget_files_gpu.ps1

Same question for generate_nuspec_for_native_nuget.py

Comment (askhade) [reply csharp/OnnxRuntime.CSharp.proj:109]: No.
The CI files use this directory to store all the downloaded artifacts from different jobs which the packaging pipeline depends on. "generate_nuspec_for_native_nuget.py" script then uses the existence of this directory to determine whether this is ADO pipeline build or local build.

This step merely copies the native nuget package to this "nuget-local-artifacts" dir for better discovery. After this step there are 2 copies of the native package: 1. in <native_build_dir>\{config}\{config} and 2. <native_build_dir>\{config}\{config}\nuget-local-artifacts 

We can delete this copy step altogether.

**Comment by pranavsharma** [comment]
Looks fine. I assume you've verified generating, installing and using the pkg. Are there going to be tests that exercise this?

**Review by baijumeswani — DISMISSED** [review]
Training related updates look good.

**Review by skottmckay — APPROVED** [review]
:shipit:

File: csharp/test/Microsoft.ML.OnnxRuntime.Tests.NetCoreApp/Microsoft.ML.OnnxRuntime.Tests.NetCoreApp.csproj [hunk 2]
```diff
@@ -95,12 +103,12 @@
...
     </Compile>
-  </ItemGroup>
-
-  <ItemGroup Condition="'$(OnDeviceTraining)' == 'true'">
     <Compile Include="..\Microsoft.ML.OnnxRuntime.Tests.Common\TrainingTest.cs">
       <Link>TrainingTest.cs</Link>
     </Compile>
+  </ItemGroup>
+
+  <ItemGroup Condition="'$(TrainingEnabledNativeBuild)' == 'true'">
```
Comment (pengwa) [inline csharp/test/Microsoft.ML.OnnxRuntime.Tests.NetCoreApp/Microsoft.ML.OnnxRuntime.Tests.NetCoreApp.csproj:111]: I may misunderstand it :). does this mean, if enable_training is enabled, training_api related will also be built? so is there any need we have enable_training_on_device?

File: tools/ci_build/github/azure-pipelines/templates/win-ci-vs-2019.yml [hunk 3]
```diff
@@ -181,7 +181,7 @@ jobs:
       solution: '$(Build.SourcesDirectory)\csharp\OnnxRuntime.CSharp.sln'
       configuration: '${{ parameters.BuildConfig }}'
       platform: 'Any CPU'
-      msbuildArguments: '-p:OnnxRuntimeBuildDirectory="$(Build.BinariesDirectory)" -p:OrtPackageId=$(OrtPackageId) -p:DefineConstants=USE_${{ parameters.ORT_EP_NAME }}'
```
Comment (snnn) [inline tools/ci_build/github/azure-pipelines/templates/win-ci-vs-2019.yml:184]: If you remove this part, The EP specific tests will not run.

Comment (skottmckay) [reply tools/ci_build/github/azure-pipelines/templates/win-ci-vs-2019.yml:184]: If you use `--build_csharp` (we do on line 145), I believe onnxruntime_csharp.cmake will update csharp/Directory.Build.props, which sets which EPs are enabled via  'USE_xxx' defines.  

This is a better setup as it a) allows an arbitrary combination of EPs to be used and b) doesn't use a `/p` property which is a global value that overrides any values set in csproj files.

Comment (snnn) [reply tools/ci_build/github/azure-pipelines/templates/win-ci-vs-2019.yml:184]: Got it. Thank you!

File: csharp/test/Microsoft.ML.OnnxRuntime.Tests.NetCoreApp/Microsoft.ML.OnnxRuntime.Tests.NetCoreApp.csproj (hunk 2 above)
Comment (askhade) [reply csharp/test/Microsoft.ML.OnnxRuntime.Tests.NetCoreApp/Microsoft.ML.OnnxRuntime.Tests.NetCoreApp.csproj]: This is the test project. Here we use "TrainingEnabledNativeBuild" to determine whether the native ort build includes training apis and use this to determine which tests to run.

**Comment by snnn** [comment]
Windows GPU CUDA build job says:
```
[xUnit.net 00:16:41.09]     TestGpu [SKIP]
  Skipped TestGpu [1 ms
```

It doesn't seem right.

**Review by snnn — APPROVED** [review]
For your information, right now we have some problems on running CUDA tests in C#. @mszhanyi can provide more information. Right now we disabled the tests. So your change is not tested on that part.

**Review by snnn — APPROVED** [review]

**Pull request merged**